### PR TITLE
feat(github-action): Implement cfn-guard as a Github Action in TS

### PR DESCRIPTION
**Pull request by dannyvassallo**
*Issue #, if available:* N/A

*Description of changes:*

Utilizes the cfn-guard TS lib inside of a github action.

File changes primarily exist under `action/src`, `action/__tests/`, `.github/workflows`, and `action.yml` on the root.

If you want to read the changes like a compiler, start at `action/src/main.ts`

THIS DOES NOT INCLUDE FURTHER ATTRIBUTION AND DOES NOT INCLUDE README UPDATES

---
*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: action/__tests__/__mocks/mockSarif.ts
Comment (joshfried-aws) [inline action/__tests__/__mocks/mockSarif.ts:1]: nit: this seems to be a fixture, not a mock. 

Functionally this doesnt impact anything obviously

File: action/__tests__/main.test.ts [hunk 1]
```diff
@@ -1,89 +1,240 @@
...
+      createReview: false,
+      dataPath: 'stub',
+      rulesPath: 'stub',
+      token: 'stub'
+    })
+
+    await run()
+
+    expect(core.setFailed).toHaveBeenCalledWith(
+      'Validation failure. CFN Guard found violations.'
```
Comment (joshfried-aws) [inline action/__tests__/main.test.ts:185]: nit: we refer to it as `cfn-guard` when abbreviated

**Review by joshfried-aws — APPROVED** [review]
Asside from some nits, this looks good to me. 

I think we should add some lints. Personally I think we should enforce `;` at the end of statements. I personally find it much easier to read with them.

File: action/src/main.ts [hunk 2]
```diff
@@ -1,26 +1,52 @@
...
+import { checkoutRepository } from './checkoutRepository'
+import { uploadCodeScan } from './uploadCodeScan'
+import { handleValidate } from './handleValidate'
+import { handlePullRequestRun } from './handlePullRequestRun'
+import { handlePushRun } from './handlePushRun'
+import { handleWriteActionSummary } from './handleWriteActionSummary'
+import getConfig from './getConfig'
+
+enum RunStrings {
+  ValidationFailed = 'Validation failure. CFN Guard found violations.',
```
Comment (joshfried-aws) [inline action/src/main.ts:12]: nit: cfn-guard

**Comment by dannyvassallo** [comment]
> Asside from some nits, this looks good to me.
> 
> I think we should add some lints. Personally I think we should enforce `;` at the end of statements. I personally find it much easier to read with them.

Added semi rule with autoformatting.

**Comment by benbridts** [comment]
Hey folks, I think this is awesome (looking very much forward to using it)!

One thing that would be interesting to consider (that's kinda independent of this PR), is if this should (also?) exist under https://github.com/aws-actions.
The main reason for that would be organisations allow-listing actions that can be used in their org. I think in practice it won't be a big hurdle to get both aws-actions and aws-cloudformation in an allow list, but for places where aws-actions is already allowed it's one step less.

**Comment by dannyvassallo** [comment]
> Hey folks, I think this is awesome (looking very much forward to using it)!
> 
> One thing that would be interesting to consider (that's kinda independent of this PR), is if this should (also?) exist under https://github.com/aws-actions. The main reason for that would be organisations allow-listing actions that can be used in their org. I think in practice it won't be a big hurdle to get both aws-actions and aws-cloudformation in an allow list, but for places where aws-actions is already allowed it's one step less.

@benbridts Hey! Great suggestion. During design we'd considered having the action live in it's own repo outside of the guard monorepo but ultimately decided on nesting it here. It's something to talk about though so while we have it here at least for now to do some testing, I'm bringing this back to the team to discuss. Thank you!

File: action/src/uploadCodeScan.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import { context, getOctokit } from '@actions/github';
+import { SarifReport } from 'cfn-guard';
+import { compressAndEncode } from './compressAndEncode';
+import getConfig from './getConfig';
+enum Endpoints {
+  CodeScan = 'POST /repos/{owner}/{repo}/code-scanning/sarifs'
```
Comment (SG622) [inline action/src/uploadCodeScan.ts:6]: I dunno man is using enum here is weird lol

File: action/src/uploadCodeScan.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * Uploads the SARIF report to the GitHub Code Scanning API.
+ * @param {UploadCodeScanParams} params - The parameters for the code scan upload.
+ * @returns {Promise<void>} - Resolves when the code scan has been uploaded successfully.
+ */
+export async function uploadCodeScan({
+  result
+}: UploadCodeScanParams): Promise<void> {
+  const { token } = getConfig();
+  const ref = context.payload.ref;
```
Comment (SG622) [inline action/src/uploadCodeScan.ts:22]: nit: deconstruct the payload here for ref and head_commit

File: action/src/main.ts [hunk 5]
```diff
@@ -1,26 +1,52 @@
...
 
-    // Set outputs for other workflow steps to use
-    core.setOutput('time', new Date().toTimeString())
+    if (sarifRun.results.length) {
+      if (analyze) {
+        core.setFailed(RunStrings.ValidationFailed);
+        await uploadCodeScan({ result });
+      } else {
+        const results =
+          eventName === 'pull_request'
```
Comment (SG622) [inline action/src/main.ts:38]: nit: not sure if this type will used more but let's avoid the hardcode of 'pull-request' freeze it or enum it

File: action/src/main.ts [hunk 6]
```diff
@@ -1,26 +1,52 @@
-import * as core from '@actions/core'
-import { wait } from './wait'
+import * as core from '@actions/core';
```
Comment (SG622) [inline action/src/main.ts:1]: not gonna cry about relative imports lol

File: action/src/handlePushRun.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ * @param {HandlePushRunParams} params - The parameters for the push run.
+ * @param {SarifRun} params.run - The SARIF run object containing the validation results.
+ * @returns {Promise<string[][]>} - An array of arrays, where each inner array represents a violation with the following format: [file path, violation message, rule ID].
+ */
+export async function handlePushRun({
+  run
+}: HandlePushRunParams): Promise<string[][]> {
+  return run.results.map(
+    ({ locations: [location], ruleId, message: { text } }) => [
+      `❌ ${location.physicalLocation.artifactLocation.uri}:L${location.physicalLocation.region.startLine},C${location.physicalLocation.region.startColumn}`,
```
Comment (SG622) [inline action/src/handlePushRun.ts:18]: nit: optional chaining?

File: action/src/handlePullRequestRun.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
+import { context, getOctokit } from '@actions/github';
+import { SarifRun } from 'cfn-guard';
+import getConfig from './getConfig';
+
+enum HandlePullRequestRunStrings {
+  Error = 'Tried to handle pull request result but could not find PR context.'
```
Comment (SG622) [inline action/src/handlePullRequestRun.ts:6]: I wonder if we should have a central place for error strings. Seems this pattern is used more .

**Review by SG622 — APPROVED** [review]

File: action/src/main.ts [hunk 9]
```diff
@@ -1,26 +1,52 @@
...
-    core.debug(new Date().toTimeString())
+  try {
+    const result = await handleValidate();
+    const {
+      runs: [sarifRun]
+    } = result;
 
-    // Set outputs for other workflow steps to use
-    core.setOutput('time', new Date().toTimeString())
+    if (sarifRun.results.length) {
```
[the remark this answers is not in the document]
Comment (dannyvassallo) [reply action/src/main.ts:32]: It should have a result always but when theres no violations the array is empty

File: action/src/main.ts [hunk 10]
```diff
@@ -1,26 +1,52 @@
...
+          await handleWriteActionSummary({
+            results
+          });
+        }
+      }
+    }
   } catch (error) {
-    // Fail the workflow run if an error occurs
-    if (error instanceof Error) core.setFailed(error.message)
+    core.setFailed(`${RunStrings.Error}: ${error}`);
```
[the remark this answers is not in the document]
Comment (dannyvassallo) [reply action/src/main.ts:50]: good catch!

File: action/src/main.ts [hunk 11]
```diff
@@ -1,26 +1,52 @@
...
+    const result = await handleValidate();
+    const {
+      runs: [sarifRun]
+    } = result;
 
-    // Set outputs for other workflow steps to use
-    core.setOutput('time', new Date().toTimeString())
+    if (sarifRun.results.length) {
+      if (analyze) {
+        core.setFailed(RunStrings.ValidationFailed);
```
[the remark this answers is not in the document]
Comment (dannyvassallo) [reply action/src/main.ts:34]: Nah this sets the workflow as a failure so the user knows that validation failed and then the report is uploaded to the github codeql section

File: action/src/index.ts [hunk 12]
```diff
@@ -1,7 +1,7 @@
 /**
  * The entrypoint for the action.
  */
-import { run } from './main'
+import { run } from './main';
 
 // eslint-disable-next-line @typescript-eslint/no-floating-promises
-run()
+run();
```
[the remark this answers is not in the document]
Comment (dannyvassallo) [reply action/src/index.ts:7]: This is from the boilerplate - not sure why its like this but I just left it. The diff is from adding the semicolons

File: action/src/uploadCodeScan.ts (hunk 3 above)
Comment (dannyvassallo) [reply action/src/uploadCodeScan.ts:6]: lol I just didnt want strings hardcoded - could go with consts.

File: action/src/main.ts (hunk 5 above)
Comment (dannyvassallo) [reply action/src/main.ts:38]: yeah - ill make a shared enum for event names good call

File: action/src/main.ts (hunk 6 above)
Comment (dannyvassallo) [reply action/src/main.ts:1]: yeah lolol i didn't want to deal with configuring it. could be a nice QOL update as a ff

File: action/src/handlePushRun.ts (hunk 7 above)
Comment (dannyvassallo) [reply action/src/handlePushRun.ts:18]: If there's a result there should be location data here

File: action/src/handlePullRequestRun.ts (hunk 8 above)
Comment (dannyvassallo) [reply action/src/handlePullRequestRun.ts:6]: Yeah I can make a shared strings or consts

**Pull request merged**